go/service/gregor.go

```diff
@@ -1016,36 +1043,88 @@ func (g *gregorHandler) auth(ctx context.Context, cli rpc.GenericClient) (err er
 	return nil
 }
 
+func (g *gregorHandler) isReachable() bool {
+	ctx := context.Background()
+	timeout := g.G().Env.GetGregorPingTimeout()
+	url, err := url.Parse(g.G().Env.GetGregorURI())
+	if err != nil {
+		g.Debug(ctx, "isReachable: failed to parse server uri, exiting: %s", err.Error())
+		return false
+	}
+
+	// If we currently think we are online, then make sure
+	conn, err := net.DialTimeout("tcp", url.Host, timeout)
+	if conn != nil {
+		conn.Close()
+		return true
+	}
+	if err != nil {
+		g.Debug(ctx, "isReachable: error: terminating connection: %s", err.Error())
+		g.Shutdown()
+		if err := g.Connect(g.uri); err != nil {
```

Comment:
Seems weird for reconnect being called as a side effect of isReachable() and that it would be doing "stuff" before being returned.

For example, if you return false but Connect succeeds before that it could be wrong? 

Maybe there should be a `checkReconnect` method that defers a shutdown and connect.

Unless the ordering here matters and there is no "cleaner" way.